Data/Shaders/VPT/VptUtils.glsl

```diff
@@ -855,110 +990,96 @@ bool getIsoSurfaceHit(
-        vec3 rdfNee = isoSurfaceColorDef / M_PI * thetaNee;
-#endif
-#ifdef SURFACE_BRDF_BLINN_PHONG
-        // http://www.thetenthplanet.de/archives/255
-        vec3 halfwayVectorNee = normalize(-w + dirLightNee);
-        vec3 rdfOut = isoSurfaceColorDef * (pow(max(dot(surfaceNormal, halfwayVectorOut), 0.0), n) / norm);
-        vec3 rdfNee = isoSurfaceColorDef * (pow(max(dot(surfaceNormal, halfwayVectorNee), 0.0), n) / norm);
-#endif
+        float pdfSamplingOut;
+        float pdfSamplingNee;
 
-        colorOut = rdfOut / pdfSamplingOut;
+        vec3 rdfNee = evaluateBrdfNee(
+            normalize(-w), dirOut, dirLightNee, surfaceNormal, surfaceTangent,
+            surfaceBitangent, isoSurfaceColorDef, useMIS, samplingPDF, hitFlags,
+            pdfSamplingOut, pdfSamplingNee);
 
-#ifdef USE_MIS
+#ifdef BRDF_SUPPORTS_SPECULAR
+        if (useMIS) {
+            // NEE with MIS.
+            // Power heuristic with beta=2: https://www.pbr-book.org/3ed-2018/Monte_Carlo_Integration/Importance_Sampling
+            //float weightNee = pdfLightNee * pdfLightNee / (pdfLightNee * pdfLightNee + pdfSamplingNee * pdfSamplingNee);
+            //float weightOut = pdfSamplingOut * pdfSamplingOut / (pdfLightOut * pdfLightOut + pdfSamplingOut * pdfSamplingOut);
 
-        // TODO: Add support for headlight.
+            float pdfLightOut = evaluateSkyboxPDF(dirOut);
 
-        // NEE with MIS.
-        // TODO: If dirOut is importance sampled from BRDF instead of cos term, use brdfOut and brdfNee
-        // instead of pdfSamplingOut and pdfSamplingNee.
-        // Power heuristic with beta=2: https://www.pbr-book.org/3ed-2018/Monte_Carlo_Integration/Importance_Sampling
-        //float weightNee = pdfLightNee * pdfLightNee / (pdfLightNee * pdfLightNee + pdfSamplingNee * pdfSamplingNee);
-        //float weightOut = pdfSamplingOut * pdfSamplingOut / (pdfSkyboxOut * pdfSkyboxOut + pdfSamplingOut * pdfSamplingOut);
-        float pdfSkyboxOut = evaluateSkyboxPDF(dirOut);
-        float weightNee = pdfLightNee / (pdfLightNee + pdfSamplingNee);
-        float weightOut = pdfSamplingOut / (pdfSkyboxOut + pdfSamplingOut);
-        colorNee +=
-                throughput * rdfNee * weightNee / pdfLightNee * calculateTransmittance(currentPoint + dirLightNee * 1e-4, dirLightNee)
-                * (sampleSkybox(dirLightNee) + sampleLight(dirLightNee));
-        colorNee +=
-                throughput * rdfOut * weightOut / pdfSamplingOut * calculateTransmittance(currentPoint + dirOut * 1e-4, dirOut)
-                * (sampleSkybox(dirOut) + sampleLight(dirOut));
+            float weightNee = pdfLightNee / (pdfLightNee + pdfSamplingNee);
+            float weightOut = pdfSamplingOut / (pdfLightOut + pdfSamplingOut);
+
+#ifdef USE_HEADLIGHT
+            // vec3 commonFactor = 2.0 * throughput * rdfNee;
+            if (isSamplingHeadlight) {
+                weightNee = 1.0;
+                weightOut = 1.0;
+
```

Comment:
The problem the comment described with weightOut is already solved, since it is set to it's limit, 1.0, in the case of sampling the headlight.

Reply:
I removed the unnecessary comment.